File: app/src/main/java/com/example/flutterandroid/router/RouteUtil.java
package com.example.flutterandroid.router;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import androidx.annotation.NonNull;

import com.idlefish.flutterboost.containers.BoostFlutterActivity;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class RouteUtil {

    public static boolean openFlutterPage(Context context, String url) {
        return openFlutterPage(context, url, new HashMap<>());
    }

    public static boolean openFlutterPage(Context context, String url, @NonNull Map params) {
        return openFlutterPage(context, url, params, 0);
    }

    public static boolean openFlutterPage(Context context, String url, @NonNull Map params, int requestCode) {
        try {
            Intent intent = BoostFlutterActivity.withNewEngine().url(url).params(params)
                    .backgroundMode(BoostFlutterActivity.BackgroundMode.opaque).build(context);
            if (context instanceof Activity) {
                Activity activity = (Activity) context;
                activity.startActivityForResult(intent, requestCode);
            } else {
                context.startActivity(intent);
            }
            return true;
        } catch (Throwable t) {
            return false;
        }
    }

    public static boolean openNativePage(Context context, String url) {
        return openNativePage(context, url, 0);
    }

    public static boolean openNativePage(Context context, String url, int requestCode) {
        try {
            Uri uri = Uri.parse(url);
            Intent intent = new Intent();
            initParams(intent, url);
            intent.setAction(Intent.ACTION_VIEW);
            intent.setData(uri);
            if (context instanceof Activity) {
                Activity activity = (Activity) context;
                activity.startActivityForResult(intent, requestCode);
            } else {
                context.startActivity(intent);
            }
            return true;
        } catch (Throwable t) {
            return false;
        }
    }

    private static void initParams(Intent intent, String url) {
        Uri uri = Uri.parse(url);
        Set<String> params = uri.getQueryParameterNames();
        if (params == null) {
            return;
        }
        String v;
        for (String name : params) {
            v = uri.getQueryParameter(name);
            if ("true".equals(v)) {
                intent.putExtra(name, true);
            } else if ("false".equals(v)) {
                intent.putExtra(name, false);
            } else {
                try {
                    int dv = Integer.parseInt(v);
                    intent.putExtra(name, dv);
                } catch (Throwable e) {
                    intent.putExtra(name, v);
                }
            }
        }
    }

}

File: app/src/main/java/com/example/flutterandroid/router/RouteUtil.kt
package com.example.flutterandroid.router

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import com.idlefish.flutterboost.containers.BoostFlutterActivity
import java.io.Serializable
import java.util.*

object RouteUtil {

    /**
     * 打开页面
     */
    fun openPage(
        context: Context,
        url: String,
        params: Map<String, Any> = HashMap<String, Any>(),
        requestCode: Int = 0
    ): Boolean {
        return when {
            url.startsWith(RouteConstant.flutterScheme) -> {
                openFlutterPage(context, url, params, requestCode)
                true
            }
            url.startsWith(RouteConstant.nativeScheme) -> {
                openNativePage(context, url, params, requestCode)
                true
            }
            else -> {
                Log.e("RouteUtil", "${url}页面不存在")
                false
            }
        }
    }

    /**
     * 打开Flutter页面
     */
    private fun openFlutterPage(
        context: Context,
        url: String,
        params: Map<String, Any> = HashMap<String, Any>(),
        requestCode: Int = 0
    ): Boolean {
        return try {
            val intent = BoostFlutterActivity
                .withNewEngine()
                .url(url)
                .params(params)
                .backgroundMode(BoostFlutterActivity.BackgroundMode.opaque)
                .build(context)
            if (context is Activity) {
                context.startActivityForResult(intent, requestCode)
            } else {
                context.startActivity(intent)
            }
            true
        } catch (t: Throwable) {
            false
        }
    }

    /**
     * 打开Native页面
     */
    private fun openNativePage(
        context: Context,
        url: String,
        params: Map<String, Any> = HashMap<String, Any>(),
        requestCode: Int = 0
    ): Boolean {
        return try {
            val uri = Uri.parse(url)
            val intent = Intent()
            params.entries.forEach { out ->
                if (out.value is Map<*, *>) {
                    (out.value as Map<*, *>).entries.forEach { inner ->
                        if (inner.key != null && inner.value != null) {
                            putExtra(intent, inner.key.toString(), inner.value!!)
                        }
                    }
                }
            }
            intent.action = Intent.ACTION_VIEW
            intent.data = uri
            if (context is Activity) {
                context.startActivityForResult(intent, requestCode)
            } else {
                context.startActivity(intent)
            }
            true
        } catch (t: Throwable) {
            false
        }
    }

    private fun putExtra(intent: Intent, key: String, value: Any) {
        when (value) {
            is Boolean -> {
                intent.putExtra(key, value)
            }
            is Byte -> {
                intent.putExtra(key, value)
            }
            is Char -> {
                intent.putExtra(key, value)
            }
            is Short -> {
                intent.putExtra(key, value)
            }
            is Int -> {
                intent.putExtra(key, value)
            }
            is Long -> {
                intent.putExtra(key, value)
            }
            is Float -> {
                intent.putExtra(key, value)
            }
            is Double -> {
                intent.putExtra(key, value)
            }
            is String -> {
                intent.putExtra(key, value)
            }
            is CharSequence -> {
                intent.putExtra(key, value)
            }
            is Parcelable -> {
                intent.putExtra(key, value)
            }
            is Serializable -> {
                intent.putExtra(key, value)
            }
            is Bundle -> {
                intent.putExtra(key, value)
            }
        }
    }
}
